go/tools/gopackagesdriver/bazelquerydriver/driver.go

```diff
@@ -0,0 +1,323 @@
+			bazel:         bzl,
+			sdk:           sdk,
+			stdlibImports: make(map[string]bool),
+			fileQueries:   make(map[string]bool),
+			importQueries: make(map[string]bool),
+		}
+		return driver.loadPackages(patterns...)
+	}
+}
+
+func (d *bazelDriver) loadPackages(patterns ...string) (*protocol.Response, error) {
+	if info, err := d.bazel.Info(); err != nil {
+		return &protocol.Response{NotHandled: true}, nil
+	} else {
+		d.workspaceRoot = info["workspace"]
+	}
+
+	log.Printf("mode: %v", d.cfg.Mode)
+	unsupportedModes := d.cfg.Mode &^ supportedModes
+	if unsupportedModes != 0 {
+		return nil, fmt.Errorf("%v (%b) not implemented", unsupportedModes, unsupportedModes)
+	}
+
+	if len(patterns) == 0 {
+		patterns = append(patterns, ".")
+	}
+
+	var ignoredPkg packages.Package
+
+	for _, patt := range patterns {
+		switch {
+		case strings.HasPrefix(patt, fileQueryPrefix):
+			fp := strings.TrimPrefix(patt, fileQueryPrefix)
+			if err := d.processFileQuery(fp, &ignoredPkg); err != nil {
+				return nil, err
+			}
+		case patt == ".":
+			d.bazelQueries = append(d.bazelQueries, goFilter(":all"))
+			d.wildcardQuery = true
+		case patt == "./...":
+			d.bazelQueries = append(d.bazelQueries, goFilter("..."))
+			d.wildcardQuery = true
+		case d.sdk.packages[patt]:
+			d.stdlibImports[patt] = true
+		default:
+			// TODO: do we need to check for a package ID instead of import path?
+			d.importQueries[patt] = true
+			d.bazelQueries = append(d.bazelQueries, fmt.Sprintf("attr(importpath, '%v', deps(%v))", patt, goFilter("//...")))
+		}
+	}
+
+	var resp protocol.Response
+
+	for pkg := range d.stdlibImports {
+		resp.Roots = append(resp.Roots, pkg)
+	}
+
+	if len(d.bazelQueries) != 0 {
+		pkgs, roots, err := d.packagesFromQueries(d.bazelQueries)
+		if err != nil {
+			return nil, err
+		}
+		resp.Packages = append(resp.Packages, pkgs...)
+		resp.Roots = append(resp.Roots, roots...)
+	}
+
+	if len(ignoredPkg.IgnoredFiles) > 0 {
+		log.Printf("Ignored %v files", len(ignoredPkg.IgnoredFiles))
+		ignoredPkg.GoFiles = ignoredPkg.IgnoredFiles[:1]
+		ignoredPkg.CompiledGoFiles = ignoredPkg.GoFiles
+		ignoredPkg.IgnoredFiles = ignoredPkg.IgnoredFiles[1:]
+		resp.Packages = append(resp.Packages, &ignoredPkg)
+		resp.Roots = append(resp.Roots, ignoredPkg.ID)
+	}
+
+	stdlib, err := d.sdk.loadPackages(&d.cfg, d.stdlibImports)
+	if err != nil {
+		return nil, err
+	}
+	resp.Packages = append(resp.Packages, stdlib...)
+
+	if d.cfg.Mode&packages.NeedTypesSizes != 0 {
+		goarch := protocol.GetEnv(&d.cfg, "GOARCH", runtime.GOARCH)
+		// SizesFor always return StdSizesm,
+		resp.Sizes = types.SizesFor("gc", goarch).(*types.StdSizes)
+	}
+
+	return &resp, nil
+}
+
+func (d *bazelDriver) processFileQuery(fp string, ignoredPkg *packages.Package) error {
+	if len(fp) == 0 {
+		return fmt.Errorf("\"file=\" prefix given with no query after it")
+	}
+	stdlibRoot := filepath.Join(d.sdk.goroot, "src")
+
+	if filepath.IsAbs(fp) {
+		if strings.HasPrefix(fp, stdlibRoot) {
```

Comment:
In general, `strings.HasPrefix` isn't safe to use on paths, since there might be other characters between the prefix and the separator (like `$GOROOT/srcccc`).

Maybe copy in a function like Gazelle's [`pathtools.HasPrefix`](https://github.com/bazelbuild/bazel-gazelle/blob/3fccaeca6a77cc41adcb90c4c8ce0af5c49d2c9d/pathtools/path.go#L32) instead?

Reply:
Sure, that's fine in this case, since we don't need to match the `src` directory itself.